### PR TITLE
Feature/ios 271 database

### PR DESCRIPTION
**Pull request by DmitryBespalov**
Added Database, Connection, Statement, ResultSet and their SQLite implementations.

```
let userID = 1
let connection = try db.connection()
let statement = try connection.prepare(statement: "SELECT session_id FROM tbl_user WHERE user_id = ?")
try statement.set(1, at: 1)
guard let resultSet = try statement.execute(), resultSet.advanceToNextRow() else { return nil }
guard let session_id = resultSet.string(at: 0) else { return nil }
return try User(id: try UserID(userID), sessionID: try SessionID(session_id))
```

### CODE REVIEW

File: IdentityAccess/IdentityAccessImplementations/persistence/CSQLite3ErrorTests.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+extension CSQLite3ErrorTests {
+
+    private func assertErrorMessage(_ code: Int32) {
+        let msg = errorMessage(code)
+        if msg == nil {
+            XCTFail("No error message provided for code \(code)")
+        } else {
+            print(msg!)
```
Comment (sche) [inline IdentityAccess/IdentityAccessImplementations/persistence/CSQLite3ErrorTests.swift:111]: should it be here?

File: IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    var libversion_number_result: Int32 = 0
+
+    var open_result: Int32 = 0
+    var open_pointer_result: OpaquePointer?
+    override func sqlite3_open(_ filename: UnsafePointer<Int8>!,
+                               _ ppDb: UnsafeMutablePointer<OpaquePointer?>!) -> Int32 {
+        if let fn = filename {
+            openedFilename = String(cString: fn, encoding: .utf8)
+        }
+        ppDb.pointee = open_pointer_result
```
Comment (sche) [inline IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift:31]: what is "pointee"?

File: IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        return open_result
+    }
+
+    override func sqlite3_libversion_number() -> Int32 {
+        return libversion_number_result
+    }
+
+    override func sqlite3_libversion() -> UnsafePointer<Int8>! {
+        return libversion_result.withCString { ptr -> UnsafePointer<Int8> in ptr }
+
```
Comment (sche) [inline IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift]: empty line

File: IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        bind_text_in_destructor = destructor
+        return bind_text_result
+    }
+
+    var bind_double_result: Int32 = 0
+    var bind_double_in_pStmt: OpaquePointer?
+    var bind_double_in_index: Int32?
+    var bind_double_in_zValue: Double?
+    override func sqlite3_bind_double(_ pStmt: OpaquePointer!, _ index: Int32, _ zValue: Double) -> Int32 {
+        bind_double_in_pStmt = pStmt
```
Comment (sche) [inline IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift:168]: Do you use these vars somehow?

File: IdentityAccess/IdentityAccessImplementations/persistence/SQLiteDatabase.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+//  Copyright © 2018 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import Common
+
+public class SQLiteDatabase: Database, Assertable {
+
+    public let name: String
+    public var exists: Bool { return false }
```
Comment (sche) [inline IdentityAccess/IdentityAccessImplementations/persistence/SQLiteDatabase.swift]: should it be like this?

File: IdentityAccess/IdentityAccessImplementations/persistence/CSQLite3ErrorTests.swift (hunk 1 above)
Comment (DmitryBespalov) [reply IdentityAccess/IdentityAccessImplementations/persistence/CSQLite3ErrorTests.swift:111]: yes, that's for eyeballing the result of sqlite errorr message calls

Comment (DmitryBespalov) [reply IdentityAccess/IdentityAccessImplementations/persistence/CSQLite3ErrorTests.swift:111]: just in tests

File: IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift (hunk 2 above)
Comment (DmitryBespalov) [reply IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift:31]: pointee is the address to where pointer is pointing (target)

File: IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift (hunk 4 above)
Comment (DmitryBespalov) [reply IdentityAccess/IdentityAccessImplementations/persistence/MockCSQLite3.swift:168]: ideally that should be used in tests to verify that correct arguments passed, but I only did minimum verification on some agrs

File: IdentityAccess/IdentityAccessImplementations/persistence/SQLiteDatabase.swift (hunk 5 above)
Comment (DmitryBespalov) [reply IdentityAccess/IdentityAccessImplementations/persistence/SQLiteDatabase.swift]: no, this should be a proper implementation. Looks like I forgot about it

**Review by sche — APPROVED** [review]

**Pull request merged**